File: 剑指offer(Sword for offer)/从尾到头打印链表.py
 -*- coding:utf-8 -*-
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

# 方法一：使用insert函数对列表进行头插法
class Solution:
    # 返回从尾部到头部的列表值序列，例如[1,2,3]
    def printListFromTailToHead(self, listNode):
        # write code here
        l = []
        head = listNode
        while head:
            l.insert(0, head.val)
            head = head.next
        return l


# 方法2：用[::-1]进行逆序
class Solution:
    # 返回从尾部到头部的列表值序列，例如[1,2,3]
    def printListFromTailToHead(self, listNode):
        # write code here
        l = []
        while listNode:
            l.append(listNode.val)
            listNode = listNode.next
        return l[::-1]

File: 剑指offer(Sword for offer)/从尾到头打印链表.cpp
//输入一个链表，按链表从尾到头的顺序返回一个ArrayList。

#include<iostream>
#include<vector>
using namespace std;

struct ListNode {
      int val;
      struct ListNode *next;
      ListNode(int x) :
           val(x), next(NULL) {
     }
};

// 方法：建立两个vector，一个用于接收顺序的链表元素，一个用于倒序排列这些元素
class Solution {
public:
    vector<int> printListFromTailToHead(ListNode* head) {
        vector<int> arrayList;
        vector<int> rearrayList;
		ListNode* h = head;
		while(h != NULL){
			arrayList.push_back(h->val);
			h = h->next;
		}
        int len = arrayList.size();
        for(int i = len - 1; i >= 0; --i){
            rearrayList.push_back(arrayList[i]);
        }
		return rearrayList;
    }
};
